src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "kafka/server/handlers/handler_probe.h"
+
+#include "config/configuration.h"
+#include "kafka/server/handlers/handler_interface.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/metrics.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace kafka {
+
+handler_probe_manager::handler_probe_manager()
+  : _probes(max_api_key() + 2)
+  , _unknown_handler_index(max_api_key() + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key)
+  , _last_recorded_in_progress(ss::lowres_clock::now()) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka_handler"),
+      {
+        sm::make_counter(
+          "Kafka_requests_completed_per_handler",
+          [this] { return _requests_completed; },
+          sm::description("Number of requests completed"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "kafka_requests_errored_per_handler",
+          [this] { return _requests_errored; },
+          sm::description("Number of requests errored"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "kafka_requests_in_progress_per_handler",
+          [this] { return _requests_in_progress_every_ns / 1'000'000'000; },
+          sm::description("An integral over requests in progress"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "kafka_bytes_received_per_handler",
```

Comment:
Units should be at the end
```suggestion
          "received_bytes",

Reply:
Will reorder. Wondering if `kafka_per_handler_received_bytes_total` would be the best way to name it.

Reply:
I don't know why we'd want `per_handler`; all metrics are implicitly "per label set", each label value combination (and the metric name is the special label `__name__`) is a time series. 

Some hints are here: https://prometheus.io/docs/practices/naming/

Reply:
@travisdowns 
I don't believe it is a breaking change. If you add a new label and don't specify that label in the PromQL, then it will avg across all the values of that label, which should be the same result.

Reply:
@travisdowns It won't conflict with any existing metrics as the group name is unique. I.e., all the metrics will be prepended with `kafka_handler_*` in the raw results we return.